### PR TITLE
Sindhu/load ckpt

### PR DESCRIPTION
**Pull request by sindhu-nervana**

### CODE REVIEW

File: diagnostics/model_test/verify_model.py [hunk 1]
```diff
@@ -67,41 +67,72 @@ def calculate_output(param_dict, select_device, input_example):
...
 
-    graph_def = tf.GraphDef()
-    if graph_filename.endswith("pbtxt"):
-        with open(graph_filename, "r") as f:
-            text_format.Merge(f.read(), graph_def)
-    else:
-        with open(graph_filename, "rb") as f:
-            graph_def.ParseFromString(f.read())
+    config = tf.ConfigProto(
+        allow_soft_placement=True,
```
Comment (avijit-nervana) [inline diagnostics/model_test/verify_model.py]: Why do you need soft placement=true?

Comment (sindhu-nervana) [reply diagnostics/model_test/verify_model.py]: @avijit-nervana Hmm, it is not needed, will remove it. Thanks for pointing out.

File: diagnostics/model_test/verify_model.py [hunk 2]
```diff
@@ -61,47 +61,85 @@ def calculate_output(param_dict, select_device, input_example):
...
-    graph_def = tf.GraphDef()
-    if graph_filename.endswith("pbtxt"):
-        with open(graph_filename, "r") as f:
-            text_format.Merge(f.read(), graph_def)
-    else:
-        with open(graph_filename, "rb") as f:
-            graph_def.ParseFromString(f.read())
+    sess = tf.Session(config=config)
+    import pdb
+    #pdb.set_trace()
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: `pdb`s can be removed

File: diagnostics/model_test/verify_model.py [hunk 3]
```diff
@@ -61,47 +61,85 @@ def calculate_output(param_dict, select_device, input_example):
...
+        is_ckpt = True
+    if tf.gfile.Exists(
+            graph_filename) and not tf.gfile.Exists(checkpoint_filename):
+        is_ckpt = False
+    if not tf.train.checkpoint_exists(
+            checkpoint_filename) and not tf.gfile.Exists(graph_filename):
+        raise Exception("Input graph file '" + graph_filename + "' or" +
+                        "Input checkpoint file '" + checkpoint_filename +
+                        "' does not exist!")
+    #pdb.set_trace()
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: `pdb` here as well

File: diagnostics/model_test/verify_model.py [hunk 4]
```diff
@@ -61,47 +61,85 @@ def calculate_output(param_dict, select_device, input_example):
...
+            checkpoint_filename) and not tf.gfile.Exists(graph_filename):
+        is_ckpt = True
+    if tf.gfile.Exists(
+            graph_filename) and not tf.gfile.Exists(checkpoint_filename):
+        is_ckpt = False
+    if not tf.train.checkpoint_exists(
+            checkpoint_filename) and not tf.gfile.Exists(graph_filename):
+        raise Exception("Input graph file '" + graph_filename + "' or" +
+                        "Input checkpoint file '" + checkpoint_filename +
+                        "' does not exist!")
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: please refactor them into if-elses.

```
chkpt_exists = tf.train.checkpoint_exists(checkpoint_filename)
graphfile_exists = tf.gfile.Exists(graph_filename)
if (chkpt_exists == graphfile_exists): #XNOR
    raise error #both cant be present and both cant be absent
```

File: diagnostics/model_test/verify_model.py [hunk 5]
```diff
@@ -61,47 +61,85 @@ def calculate_output(param_dict, select_device, input_example):
...
         if len(output_tensor_name) == 0:
             # if no outputs are specified, then compare for all tensors
             output_tensor_name = sum(
-                [[j.name for j in i.outputs] for i in graph.get_operations()],
-                [])
+                [[j.name
+                  for j in i.outputs]
+                 for i in graph_from_ckpt.get_operations()], [])
+
+    if tf.gfile.Exists(graph_filename):
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: put this in the else portion of `is_ckpt`

File: diagnostics/model_test/verify_model.py [hunk 6]
```diff
@@ -61,47 +61,85 @@ def calculate_output(param_dict, select_device, input_example):
...
+            with open(graph_filename, "rb") as f:
+                graph_def.ParseFromString(f.read())
+
+        with tf.Graph().as_default() as graph:
+            tf.import_graph_def(graph_def)
+            if len(output_tensor_name) == 0:
+                # if no outputs are specified, then compare for all tensors
+                output_tensor_name = sum([
+                    [j.name for j in i.outputs] for i in graph.get_operations()
+                ], [])
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: Is there any difference between `graph` and  `graph_from_ckpt`?
If not please refactor the `output_tensor_name` computation outside the if-else

File: diagnostics/model_test/verify_model.py [hunk 7]
```diff
@@ -61,47 +61,85 @@ def calculate_output(param_dict, select_device, input_example):
...
 
     # Create the tensor to its corresponding example map
     tensor_to_example_map = {}
     for item in input_example:
-        t = graph.get_tensor_by_name(item)
+        if (is_ckpt):
+            t = graph_from_ckpt.get_tensor_by_name(item)
+        else:
+            t = graph.get_tensor_by_name(item)
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: same comment. `graph_from_ckpt` and `graph` are probably same, so no need for an if-else here

File: diagnostics/model_test/verify_model.py [hunk 8]
```diff
@@ -61,47 +61,85 @@ def calculate_output(param_dict, select_device, input_example):
...
+        output = sess.run(output_tensor, feed_dict=tensor_to_example_map)
+        return output, output_tensor_name
+    else:
+        output_tensor = [
+            graph.get_tensor_by_name(i) for i in output_tensor_name
+        ]
+        with tf.Session(graph=graph, config=config) as sess_pb:
+            output_tensor = sess_pb.run(
+                output_tensor, feed_dict=tensor_to_example_map)
+            return output_tensor, output_tensor_name
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: This if-else is not required. This code can be common for ckpt, pb and pbtxt

File: diagnostics/model_test/verify_model.py (hunk 4 above)
Comment (sindhu-nervana) [reply diagnostics/model_test/verify_model.py]: working on adding/modifying some if-conditions only, hence WIP label :)

File: diagnostics/model_test/verify_model.py [hunk 9]
```diff
@@ -60,30 +60,63 @@ def calculate_output(param_dict, select_device, input_example):
...
 
-    graph_def = tf.GraphDef()
-    if graph_filename.endswith("pbtxt"):
-        with open(graph_filename, "r") as f:
-            text_format.Merge(f.read(), graph_def)
     else:
-        with open(graph_filename, "rb") as f:
-            graph_def.ParseFromString(f.read())
+        raise Exception(
+            "Input graph file OR Input checkpoint file is required!")
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: this will not raise an error if both are present.

**Comment by sindhu-nervana** [comment]
TESTNOW

File: diagnostics/model_test/verify_model.py [hunk 10]
```diff
@@ -60,30 +60,63 @@ def calculate_output(param_dict, select_device, input_example):
...
+
+    output_tensor_name = param_dict["output_tensor_name"]
+
+    config = tf.ConfigProto(
+        # log_device_placement=True,
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: this can be removed

File: diagnostics/model_test/verify_model.py [hunk 11]
```diff
@@ -60,30 +60,63 @@ def calculate_output(param_dict, select_device, input_example):
...
+
+        if not tf.train.checkpoint_exists(checkpoint_filename):
+            raise Exception("Checkpoint with this prefix does not exist")
+        else:
+            saver.restore(sess, checkpoint_filename)
+
+        print("Model restored.")
+        graph = tf.get_default_graph()
+
+    #if graph,then load graph
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: minor:
`graph,then` -> `graph, then`

File: diagnostics/model_test/template.json [hunk 12]
```diff
@@ -2,13 +2,14 @@
     "model_name": "name_of_your_model",
     "reference_backend": "your_reference_backend",
     "testing_backend": "your_testing_backend",
-    "frozen_graph_location": "/path/to/the/frozen_graph.pb",
+    "graph_location": "/path/to/the/frozen_graph.pb",
```
Comment (mingshan-wang) [inline diagnostics/model_test/template.json]: Maybe provide two template.jsons, one is template_pb.json and template_ckpt.json. I feel like it will be confusing if the template itself has two fields at the same time.

Comment (sindhu-nervana) [reply diagnostics/model_test/template.json]: good point, will do that.

File: diagnostics/model_test/README.md [hunk 13]
```diff
@@ -1,17 +1,17 @@
 # Compare model output between two different backends
 
-This model_test tool will run the model inference seperately on any two specified backends(e.g. Backends: Tensorflow (native), nGraph-CPU, nGraph-GPU etc) in json file and the outputs from each backend should match given the same inputs. It can be used as a debugging tool for layer by layer comparison, and also a verification that nGraph produces the same output as Tensorflow. 
+This model_test tool will run the model inference and training seperately on any two specified backends(e.g. Backends: Tensorflow (native), nGraph-CPU, nGraph-GPU etc) in json file and the outputs from each backend should match given the same inputs. It can be used as a debugging tool for layer by layer comparison, and also a verification that nGraph produces the same output as Tensorflow.
```
[the remark this answers is not in the document]
Comment (sindhu-nervana) [reply diagnostics/model_test/README.md]: @shresthamalik Thank you for pointing this out. I have verified mnist_mlp training with multiple output tensors and verified. I modified the .json file with the same and README.

**Comment by sindhu-nervana** [comment]
@sayantan-nervana @shresthamalik Could you please review this PR when you get a chance. Thanks!

File: diagnostics/model_test/verify_model.py [hunk 14]
```diff
@@ -134,12 +169,18 @@ def calculate_norm(ngraph_output, tf_output, desired_norm):
...
         raise Exception('Only L2, L2, and inf norms are supported')
 
-    n = np.linalg.norm((ngraph_output_flatten - tf_output_flatten),
-                       desired_norm)
-    if desired_norm is np.inf:
-        return n
-    else:
-        return n / len(ngraph_output_flatten)
+    if ngraph_output_flatten.size is not 0:
+        try:
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: Please add why we try-except here. I think to catch errors in type-conversion?

File: diagnostics/model_test/README.md [hunk 15]
```diff
@@ -1,23 +1,25 @@
...
+This model_test tool will run the model seperately on any two specified backends(e.g. Backends: Tensorflow (native), nGraph-CPU, nGraph-GPU etc) in json file and the outputs from each backend should match given the same inputs. It can be used as a debugging tool for layer by layer comparison, and also a verification that nGraph produces the same output as Tensorflow.  Depending on the input model specified in the .pb, .pbtxt or .ckpt it can be used to verify layers in forward pass and backward pass both.
 
 # Required files to use the tool:
-* A json file: Provide model specific parameters. Look at the example ```mnist_cnn.json```. You can start with the ```template.json``` and modify it to match your model
-* A tensorflow model file: Either .pb or .pbtxt file
+* A json file: Provide model specific parameters. Look at the example ```mnist_cnn.json``` to use a .pb or .pbtxt graph and ```mnist_mlp.json``` for checkpoint graph
+* For inference verification,a tensorflow model inference .pb or .pbtxt graph file is required
+* For training verification, a tensorflow model training checkpoint(.ckpt)graph file is required
```
Comment (shresthamalik) [inline diagnostics/model_test/README.md]: You can mention here (.ckpt + .meta)

File: diagnostics/model_test/verify_model.py [hunk 16]
```diff
@@ -208,12 +252,19 @@ def parse_json():
...
-    result_tf_graph_arrs, out_tensor_names_cpu = calculate_output(
+    result_tf_graph_arrs, out_tensor_names_cpu, tf_skipped_tensors = calculate_output(
         parameters, device1, input_tensor_dim_map)
     # Run the model on testing backend
-    result_ngraph_arrs, out_tensor_names_ngraph = calculate_output(
+    result_ngraph_arrs, out_tensor_names_ngraph, ngraph_skipped_tensors = calculate_output(
         parameters, device2, input_tensor_dim_map)
 
+    assert all(
+        [i == j for i, j in zip(tf_skipped_tensors, ngraph_skipped_tensors)])
```
Comment (sayantan-nervana) [inline diagnostics/model_test/verify_model.py]: Zipping two lists will zip upto the shorter list. So we might need something like:
```
assert (len(tf_skipped_tensors) == len(ngraph_skipped_tensors)) and all(
        [i == j for i, j in zip(tf_skipped_tensors, ngraph_skipped_tensors)])
```

File: diagnostics/model_test/README.md [hunk 17]
```diff
@@ -1,23 +1,25 @@
...
+This model_test tool will run the model seperately on any two specified backends(e.g. Backends: Tensorflow (native), nGraph-CPU, nGraph-GPU etc) in json file and the outputs from each backend should match given the same inputs. It can be used as a debugging tool for layer by layer comparison, and also a verification that nGraph produces the same output as Tensorflow.  Depending on the input model specified in the .pb, .pbtxt or .ckpt it can be used to verify layers in forward pass and backward pass both.
 
 # Required files to use the tool:
-* A json file: Provide model specific parameters. Look at the example ```mnist_cnn.json```. You can start with the ```template.json``` and modify it to match your model
-* A tensorflow model file: Either .pb or .pbtxt file
+* A json file: Provide model specific parameters. Look at the example ```mnist_cnn.json``` to use a .pb or .pbtxt graph and ```mnist_mlp.json``` for checkpoint graph
+* For inference verification,a tensorflow model inference .pb or .pbtxt graph file is required
```
Comment (sayantan-nervana) [inline diagnostics/model_test/README.md]:
```suggestion
* For inference verification, a tensorflow model inference .pb or .pbtxt graph file is required
```

Comment (sayantan-nervana) [reply diagnostics/model_test/README.md]: For inference verification, a tensorflow model inference .pb or .pbtxt graph file is required:::

Not sure about this statement. we can do inf on pb, pbtxt & ckpt, but training comparision only on ckpt.

File: diagnostics/model_test/README.md [hunk 18]
```diff
@@ -1,23 +1,25 @@
...
+This model_test tool will run the model seperately on any two specified backends(e.g. Backends: Tensorflow (native), nGraph-CPU, nGraph-GPU etc) in json file and the outputs from each backend should match given the same inputs. It can be used as a debugging tool for layer by layer comparison, and also a verification that nGraph produces the same output as Tensorflow.  Depending on the input model specified in the .pb, .pbtxt or .ckpt it can be used to verify layers in forward pass and backward pass both.
 
 # Required files to use the tool:
-* A json file: Provide model specific parameters. Look at the example ```mnist_cnn.json```. You can start with the ```template.json``` and modify it to match your model
-* A tensorflow model file: Either .pb or .pbtxt file
+* A json file: Provide model specific parameters. Look at the example ```mnist_cnn.json``` to use a .pb or .pbtxt graph and ```mnist_mlp.json``` for checkpoint graph
+* For inference comparison on TF and Ngraph, a .pb, .pbtxt or .ckpt graph file is required
+* For training comparison on TF and Ngraph,a checkpoint(.ckpt) graph file and meta graph file (.meta) is required
```
Comment (sayantan-nervana) [inline diagnostics/model_test/README.md]:
```suggestion
* For training comparison on TF and Ngraph, a checkpoint(.ckpt) graph file and meta graph file (.meta) is required
```
